packages/utils/src/validation/ObjectValidator.js

```diff
@@ -12,6 +12,10 @@ import { ValidationError } from '../errors';
  * if it is not valid
  */
 export class ObjectValidator {
+  /**
+   * @param {Object<string, Function>} fieldValidators
+   * @param {Object<string, Function>} [defaultValidators]
+   */
```

Comment:
Remember the discussion we had about deducing types from existing JS code? I'm now of the opinion we should try to do this with `jsDoc`, if possible, since this way types are defined where they are actually used. This should only be required for public members whose type cannot be deduced correctly, so it's not like we'll have to do this all the time.

The next option is defining additional types.d.ts files that extend/override the automatically deduced types. The disadvantage is that types are derived from two distinct sources and in a non-consistent way.

The final option would be creating custom `index.d.ts` files per package that include all publicly exposed members, and do not use automatic type definitions. This would result in the most ideal types, but it requires more investment to get right + up to date. Personally I would be keen on maintaining those files but I'm not sure if this applies to everyone. Another caveat is that it would be harder to maintain source code mappings from the types to code definitions.

Reply:
Your reasoning seems very sound - where we're producing derived types from a non-typescript package, we should add or alter the jsdoc